doctor/lib/ocr_utils.py

```diff
@@ -0,0 +1,345 @@
+    Pdf plumber has a method to calculate the angle of text which we use here
+
+    Traditionally this is only seen in circular stamps which confuses the
+    content, or in perpendicular text of the ninth circuit courts which also
+    confuses the text.
+    """
+    if (matrix := obj.get("matrix")) is None:
+        return True
+
+    # Remove Skew
+    my_char_ctm = CTM(*matrix)
+    if my_char_ctm.skew_x != 0:
+        return False
+    return True
+
+
+def get_page_text(page: pdfplumber.PDF.pages, strip_margin: bool) -> str:
+    """Extract page text
+
+    Using pdf plumber extract out the text of the document that is not
+    skewed (ie a stamp of approval) and extract out text removing blue text
+
+    :param page: pdf plumber page
+    :param strip_margin: a flag to crop out the margin of a document and skewed content
+    :return: text
+    """
+    if strip_margin:
+        # Crop margins and remove skewed text
+        _, _, width, height = page.bbox
+        bbox = (
+            (1 / 8.5 * width),
+            (1 / 11 * height),
+            (7.5 / 8.5 * width),
+            (10 / 11 * height),
+        )
+        doc_text = (
+            page.crop(bbox)
+            .filter(deskew)
+            .extract_text(
+                layout=True, keep_blank_chars=True, y_tolerance=5, y_density=25
+            )
+        )
+    else:
+        doc_text = page.extract_text(
+            layout=True, keep_blank_chars=True, y_tolerance=5, y_density=25
+        )
+    return doc_text
+
+
+def page_images(page: pdfplumber.pdf.Page) -> bool:
+    """Does the page have images of a certain size
+
+    Meant to exclude images that might be lines
+    """
+    for img in page.images:
+        if (
+            img.get("width") / page.width * img.get("height") / page.height
+            > 0.1
+        ) or img.get("width") * img.get("width") > 10:
+            return True
+    return False
+
+
+def page_annotations(page: pdfplumber.pdf.Page) -> bool:
+    """Does the page have annotations which could contain text
+
+    :param page: pdf plumber
+    """
+    if page.annots:
+        anno_types = [
+            str(annot.get("data").get("Subtype")) for annot in page.annots
+        ]
+        if "/'FreeText'" in anno_types or "/'Widget'" in anno_types:
+            return True
+    return False
+
+
+def find_average_char_width(block_data: pd.Series) -> int:
+    """Average character width for a block of text
+
+    :param block_data:
+    :return: average character width
+    """
+    fd = block_data[block_data.text.str.len() > 0]
+    return (fd.width / fd.text.str.len()).mean()
+
+
+def validate_ocr_text(row: pd.Series, img: Image) -> pd.Series:
+    """Validate the OCR results
+
+    Take a look at our OCR and remove the bad results
+    and if possible reprocess words one by one
+
+    :param row: row of ocr results
+    :param img: Page image
+    :return: Updated row if necessary
+    """
+    if row["left"] < 370 and row["conf"] <= 40:
```

Comment:
More comments in each of these heuristics would help me understand what they're doing.